x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java

```diff
@@ -240,35 +241,154 @@ public Automaton allowedActionsMatcher(String index) {
+         * @return {@code true} if-and-only-if this object is related to a data-stream, either by having a {@link #type} of
+         * {@link IndexAbstraction.Type#DATA_STREAM} or by being the backing index for a {@link #parentDataStream data-stream}.
+         */
+        public boolean isDataStreamRelated() {
+            switch (type) {
+                case DATA_STREAM:
+                    return true;
+                case CONCRETE_INDEX:
+                    return parentDataStream != null;
+                default:
+                    return false;
+            }
+        }
+
+        /**
+         * Check whether this object is covered by the provided permission {@link Group}.
+         * For indices that are part of a data-stream, this checks both the index name and the parent data-stream name.
+         * In all other cases, it checks the name of this object only.
+         */
+        public boolean checkIndex(Group group) {
+            if (parentDataStream != null) {
+                if (group.checkIndex(parentDataStream)) {
+                    return true;
                 }
-                isBackingIndex = indexAbstraction.getType() == IndexAbstraction.Type.CONCRETE_INDEX &&
-                        indexAbstraction.getParentDataStream() != null;
-                isDataStream = indexAbstraction.getType() == IndexAbstraction.Type.DATA_STREAM;
+            }
+            return group.checkIndex(name);
+        }
+
+        /**
+         * @return the number of distinct objects to which this expansion refers.
+         */
+        public int size() {
+            if (type == IndexAbstraction.Type.CONCRETE_INDEX) {
+                return 1;
             } else {
-                isBackingIndex = isDataStream = false;
+                return 1 + concreteIndices.size();
             }
+        }
+    }
+
+    /**
+     * Authorizes the provided action against the provided indices, given the current cluster metadata
+     */
+    public Map<String, IndicesAccessControl.IndexAccessControl> authorize(
+        String action,
+        Set<String> requestedIndicesOrAliases,
+        Map<String, IndexAbstraction> lookup,
+        FieldPermissionsCache fieldPermissionsCache
+    ) {
+
+        final List<IndexResource> resources = new ArrayList<>(requestedIndicesOrAliases.size());
+        int totalResourceCount = 0;
+
+        for (String indexOrAlias : requestedIndicesOrAliases) {
+            final IndexResource resource = IndexResource.create(indexOrAlias, lookup.get(indexOrAlias));
+            resources.add(resource);
+            totalResourceCount += resource.size();
+        }
+
+        // now... every index that is associated with the request, must be granted
+        // by at least one indices permission group
+        final Map<String, Set<FieldPermissions>> fieldPermissionsByIndex = new HashMap<>(totalResourceCount);
+        final Map<String, DocumentLevelPermissions> roleQueriesByIndex = new HashMap<>(totalResourceCount);
```

Comment:
Unlike `grantedBuilder`, these two Maps may not be populated for every concrete index because they are only calculated if there is at least one group grants the permission. So we could be allocating more than needed. But when there are indices rejected by all permission groups, maybe it is always the unhappy path (request rejected) and we don't really care that much about a bit more memory consumption?